### PR TITLE
correctly handling media delete actions in the media browser

### LINKED ISSUE #6576: Media Library: allow uploading items to be canceled/deleted
### Expected behavior
When in the Media Browser, deleting an item being uploaded should be possible, and it should also cancel its upload, 
Also if the upload is attached to a given Post, a warning should be given so the user can go edit the Post and delete it from there.

### Actual behavior
The item is not deleted, and continues to be uploaded, showing a Toast saying `wait until upload completes`

### Steps to reproduce the behavior
- start uploading a video from the Media Browser
- then long press and delete it
- observe it continues to be there, uploading, and the toast appears.

### PR DESCRIPTION
**Pull request by mzorz**
Fixes #6576 

**To test:** 
**CASE A**: uploading media item in Post can't be deleted/stopped from the Media Browser
1. start a new draft
2. include a media item (something large, like a video, so it's easier to test)
3. exit the editor, the upload will continue in the background
4. go to the Media Browser
5. you'll see the media item being uploaded
6. long press on it to select it
7. tap on the trash icon
8. when prompted to confirm deletion, tap OK
9. a Toast is shown `Some media can\'t be deleted at this time. Wait until upload completes, or try deleting it from within the associated Post.`.
10. Confirm this happens each time you try, successfully preventing the user from deleting a media item that is being uploaded as part of a Post.

**CASE B:** upload a media item from the Media Browser and delete it
1. go to the media browser
2. include a media item by tapping on the `+` button on the top-right corner of the screen
3. while the item is uploading, long press on it
4. tap on the trash icon
5. when prompted to confirm deletion, tap OK
6. verify the item is deleted and the upload is canceled (check upload progress logs for that).

Please note this doesn't cover the case of deleting media that has already been uploaded to the server and is included / referenced in some Post content, which is a separate case.
This only covers the case of being able to delete solo items, or hinting the user on what to do if they want to cancel an upload for an item that is included in a Post currently being uploaded.

cc @nbradbury @daniloercoli

### CODE REVIEW

**Comment by nbradbury** [comment]
For Case A, why do we not simply delete the media? It seems like we should do that rather than show the user a toast saying they have to delete it from the editor instead.

File: WordPress/src/main/res/values/strings.xml [hunk 1]
```diff
@@ -188,8 +188,7 @@
     <!-- Delete Media -->
     <string name="confirm_delete_media">Delete selected item?</string>
     <string name="confirm_delete_multi_media">Delete selected items?</string>
-    <string name="wait_until_upload_completes">Wait until upload completes</string>
-    <string name="cannot_delete_multi_media_items">Some media can\'t be deleted at this time. Try again later.</string>
+    <string name="cannot_delete_some_multi_media_items">Some media can\'t be deleted at this time. Wait until upload completes, or try deleting it from within the associated Post.</string>
```
Comment (nbradbury) [inline WordPress/src/main/res/values/strings.xml]: If we choose to keep the toast, this really needs to be much shorter - it's too verbose for a toast message.

File: WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java [hunk 2]
```diff
@@ -665,36 +665,39 @@ public void onSupportActionModeFinished(@NonNull ActionMode mode) {
...
+            if (!UploadService.isMediaBeingUploadedForAPost(mediaModel)) {
+
+                // if uploading, first issue a cancel upload command
+                if (UploadService.isPendingOrInProgressMediaUpload(mediaModel)) {
+                    MediaStore.CancelMediaPayload payload = new MediaStore.CancelMediaPayload(mSite, mediaModel, false);
+                    mDispatcher.dispatch(MediaActionBuilder.newCancelMediaUploadAction(payload));
+                }
+
                 if (mediaModel.getUploadState() != null &&
                         MediaUtils.isLocalFile(mediaModel.getUploadState().toLowerCase())) {
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: I know this isn't part of this PR, but we can drop the `toLowerCase()` here (and also the `null` check).

File: WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java [hunk 3]
```diff
@@ -665,36 +665,39 @@ public void onSupportActionModeFinished(@NonNull ActionMode mode) {
...
                     mDispatcher.dispatch(MediaActionBuilder.newRemoveMediaAction(mediaModel));
-                    sanitizedIds.add(String.valueOf(currentId));
-                    continue;
+                } else {
+                    mediaToDelete.add(mediaModel);
+                    mediaModel.setUploadState(MediaUploadState.DELETING);
+                    mDispatcher.dispatch(MediaActionBuilder.newUpdateMediaAction(mediaModel));
```
Comment (nbradbury) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: It looks like `handleMultiSelectDelete()` in the fragment - which is the only place that calls this routine - already sets the upload state. Should it be removed from there?

Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: addressed in ad6cbef

File: WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java (hunk 2 above)
Comment (mzorz) [inline WordPress/src/main/java/org/wordpress/android/ui/media/MediaBrowserActivity.java]: well it actually could be null, and it needs to be made sure to be lowercase so the implementation of `isLocalFile()` can work correctly. We might refactor this and convert to `MediaUploadState` properly first, using `MediaUploadState.fromString()`, and making `isLocalFile()` receive a `MediaUploadState` parameter type instead of a String, but maybe in another smallish PR as this is also being used like that on some other parts of the code? will do that right after this one 👍

File: WordPress/src/main/res/values/strings.xml (hunk 1 above)
Comment (mzorz) [inline WordPress/src/main/res/values/strings.xml]: agreed, removing this one

**Comment by mzorz** [comment]
> For Case A, why do we not simply delete the media? It seems like we should do that rather than show the user a toast saying they have to delete it from the editor instead.

ok we're doing that now in 3d44ecf and 9e6b4b7 - I was a bit uncomfortable with the fact that after deleting a media item in the media browser, a Post that had such an item within it would appear failed and marked for `Retry`. Then if the user taps on the image within the Editor to retry, we can't really retry because the `MediaModel` doesn't exist in the DB anymore, so we now show a different toast giving some explanation about this and what they can do as users. OTOH if we didn't remove it from the library, it would appear in the Media Browser, when all the user wanted to do is to actually `delete` it.
I think this is “enough” for what we want (that is, the user wants to delete an item that is being uploaded, we delete it for them and mark the Post so to bring their attention, and once they go to the Post if they want to retry they're given an indication about what to do), but I’m not totally satisfied with it (i.e. ideally I want the user to be able to stop an upload from the Media Browser and then retry it in the Editor at a later time, instead of being told they need to do something else).

Another possibility would be to directly eliminate the image from within the Post body, but that would mean modifying their Post content, which I don’t like at all.

The best solution I can think of is to mark it logically `deleted`, and not show it in the Media Browser, but still be able to recover from it if they tap `Retry` on the image within the associated Post. But in order to implement this, we’d also need some FluxC changes… and this all would demand a deeper thought than what this PR is aiming at.

So back to what we have, I believe the current state of the PR is good enough, what do you think? Also then, ready for another round @nbradbury 🙇

**Comment by nbradbury** [comment]
I think this works much better, but it's still not very user friendly. We can stick with it this way for now, but it would be nice if instead of showing that toast saying the user should try inserting the media again, we said something like "Media has been removed. Delete it from this post?"

Two issues still:
1. The existing toast is too wordy, how about simply "Media has been removed and can't be retried."
2. In this visual editor, if I tap to retry a deleted media I see the toast message but then the media shows in the post with a progress bar, as though it's still being uploaded.
screenshot_1503654454

**Comment by mzorz** [comment]
Nice touch @nbradbury! implemented the dialog here 2e7acb3 and here d9abdc0, also making sure the progress bar doesn't start getting updates if retry could not happen, as seen in that visual editor screenshot
Also changed the cancel/ok wording for no/yes in the  "Stop uploading?" dialog in facb1d3 as you suggested elsewhere
Ready for another round! 🙇

**Comment by mzorz** [comment]
Reverted the yes/no change in 40f3342 as it wasn't complete and another solution is on the way in #6590

**Comment by nbradbury** [comment]
Looks good! :shipit:
